### PR TITLE
Display RGB when hovering over images

### LINKED ISSUE #751: Display RGB values when hovering over images
### Displaying rgb values when hovering over images to provide imagedate

 This is a new feature and can be really interesting to implement

### where to look

http://sequencer.publiclab.org...

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site.

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by aashna27**
Fixes #751 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` and `@publiclab/is-reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

File: examples/lib/defaultHtmlStepUi.js [hunk 1]
```diff
@@ -156,6 +157,7 @@ function DefaultHtmlStepUi(_sequencer, options) {
       $(step.ui.querySelectorAll(".cal")).toggleClass("collapse");
     });
     
+    $(step.imgElement).on("mousemove", debounce(() => imageHover(step), 150));
```
Comment (HarshKhandeparkar) [inline examples/lib/defaultHtmlStepUi.js]: What does debounce do?

Comment (aashna27) [reply examples/lib/defaultHtmlStepUi.js]: it limits a fn being called too frequently , so that site performance isnt hampered

**Comment by HarshKhandeparkar** [comment]
Alsk please provide a GIF

File: examples/lib/defaultHtmlStepUi.js (hunk 1 above)
Comment (HarshKhandeparkar) [reply examples/lib/defaultHtmlStepUi.js]: Cool

**Comment by aashna27** [comment]
> 
> 
> Alsk please provide a GIF

yeah creating one

**Comment by HarshKhandeparkar** [comment]
Also instead of showing the words rgb, can you fit in an alpha value?

**Comment by aashna27** [comment]
> Looks good! I also have a (partly unnecessary) feature request. I don't know if this does it already. But does it show the value only if the user hovers for a specific amount of time. Like if a mouse enters the image and moves then nothing happens but if the mouse is steady for let's say 2s then the values are shown

yeah it does expect a steady wait, Actually this is with the tooltip, if I just display the results normally then it's amazingly quick, but with tooltip a little steadiness is needed.

**Comment by aashna27** [comment]
> 
> 
> Also instead of showing the words rgb, can you fit in an alpha value?

yeah rgba can be added, I ll do that.

**Comment by Divy123** [comment]
@HarshKhandeparkar I think adding RGB is  better than alpha as it explicitly defines that pixel better.

**Comment by HarshKhandeparkar** [comment]
@aashna27 one last thing. I don't know if this is necessary but please check if the values are right by maybe using an image which is half one color and half another color.

**Comment by HarshKhandeparkar** [comment]
@Divy123 this feature will mostly be used for debugging (and geeks?) I guess. So adding and alpha can help in debugging. What do you think?

**Comment by Divy123** [comment]
Agreed @HarshKhandeparkar I thought you were asking only for opacity.

File: examples/lib/defaultHtmlStepUi.js [hunk 2]
```diff
@@ -273,6 +275,25 @@ function DefaultHtmlStepUi(_sequencer, options) {
     }
   }
 
+  function imageHover(step){
+
+    var img = $(step.imgElement);
+
+    img.mousemove(function(e) {
```
Comment (HarshKhandeparkar) [inline examples/lib/defaultHtmlStepUi.js]: I'm a bit confused. Why is the mousemove event handler added again here if it is already added above?

File: examples/lib/defaultHtmlStepUi.js [hunk 3]
```diff
@@ -273,6 +275,25 @@ function DefaultHtmlStepUi(_sequencer, options) {
...
+    img.mousemove(function(e) { 
+      var canvas = document.createElement('canvas');
+      canvas.width = img.width();
+      canvas.height = img.height();
+      var context = canvas.getContext('2d');
+      context.drawImage(this,0,0);
+
+      var offset = $(this).offset();
+      var xPos = e.pageX - offset.left;
+      var yPos = e.pageY - offset.top;
```
Comment (aashna27) [inline examples/lib/defaultHtmlStepUi.js]: All this required an event for the position and that's why I added "mousemove" here and yess I did do it for onComplete and it works but I thought that instead on adding the code there and increasing it lets make it modular so I made a separate function.

**Comment by aashna27** [comment]
> 
> 
> How about drawing the canvas during `onComplete` step UI method? All that has to be done after that is update the tooltip. Also there will be no need to use debounce since the function will not slow down the browser that significantly. Also can the debounce cause the mouse move events to get out of sync and display wrong values since the mouse can move after an event was fired but the debounce function was not fired ?

Actually it can slow, cause although the tooltip is slow but the computations occur really fast, when I console.log() the values it was really fast so I added debounce.And for the out of sync I think the steadiness of the tooltip will help to not get out of sync.

**Comment by aashna27** [comment]
@jywarren

**Comment by jywarren** [comment]
Rebasing this. @aashna27 I see you used debounce here -- we have a growing problem with typeahead autocompletion:

https://github.com/publiclab/plots2/issues/3472

I'm wondering if you'd be able to help us out as this becomes more urgent. I think using debounce and/or other JavaScript throttling strategies can be part of the solution. Would you be willing to help out a bit?

**Comment by aashna27** [comment]
@jywarren thanks for the rebase, and yess I would have a look at it, but since I am completely unfamiliar with the codebase of plots2 I would take time, but yeah I will have a look at it. Thanks for considering me!!

**Comment by jywarren** [comment]
I believe there are code links to the necessary JavaScript segments -- thanks, we could use all the help we can get!

Just pushed it up here. 👍

**Comment by jywarren** [comment]
Hmm, what did I do wrong?

**Comment by jywarren** [comment]
Apologies! I just merged #803 -- i must've made some error here though. Do you think you could give it a try?

**Comment by aashna27** [comment]
> Apologies! I just merged #803 -- i must've made some error here though. Do you think you could give it a try?

No problem.. 😄

**Comment by aashna27** [comment]
I think I ll close this one , I have tried multiple times, its not building.

**Pull request closed without merging**